File: C++/547 Number of Provinces.cpp
class Solution {
public:
    int findCircleNum(vector<vector<int>>& isConnected) {
        int n = isConnected.size();
        auto g = vector<vector<int>>(n, vector<int>());
        for(int i = 0; i < n; i++) {
            for(int j = i + 1; j < n; j++) {
                if(isConnected[i][j]) {
                    g[i].push_back(j);
                    g[j].push_back(i);
                }
            }
        }

        unordered_set<int> seen;
        auto dfs = [&](this auto && dfs, int root) -> void {
            seen.insert(root);
            for(auto &e : g[root]) {
                if(seen.contains(e)) continue;
                dfs(e);
            }
        };

        int ans = 0;
        for(int i = 0; i < n; i++) {
            if(!seen.contains(i)) {
                ans++;
                dfs(i);
            }
        }
        return ans;
    }
};

File: Swift/547. Number of Provinces.swift
class Solution {
    func findCircleNum(_ isConnected: [[Int]]) -> Int {
        let n = isConnected.count
        var g: [[Int]] = .init(repeating: [], count: n)
        for i in 0 ..< n {
            for j in i + 1 ..< n {
                if isConnected[i][j] == 1 {
                    g[i].append(j)
                    g[j].append(i)
                }
            }
        }

        var seen = Set<Int>()
        func dfs(_ root: Int) {
            seen.insert(root)
            for x in g[root] {
                if !seen.contains(x) {
                    dfs(x)
                }
            }
        }

        var ans = 0
        for i in 0 ..< n {
            if !seen.contains(i) {
                ans += 1
                dfs(i)
            }
        }
        return ans
    }
}
